### PR TITLE
Adding transactions to the wf package

### PR DESCRIPTION
**Pull request by mguy-huntress**
This adds the use of transactions and tests around the new functionality.

### CODE REVIEW

File: firewall.go [hunk 1]
```diff
@@ -61,12 +98,17 @@ func New(opts *Options) (*Session, error) {
...
 		handle:     handle,
 		layerTypes: layerTypes{},
 	}
 
+	if opts.StartTransaction {
+		// Don't worry if we return an error here. We're not going to fail
+		// the creation of the session if we can't start a transaction.
+		ret.BeginTransaction(opts.TransactionFlags)
+	}
+
```
Comment (rob-no) [inline firewall.go:111]: Where's the commit transaction?

File: firewall.go [hunk 2]
```diff
@@ -89,6 +131,12 @@ func (s *Session) Close() error {
 	if s.handle == 0 {
 		return nil
 	}
+
+	// if we have a transaction in progress that was not commited, abort it now
+	if s.status.State == BeganTransaction {
+		s.AbortTransaction()
+	}
+
```
Comment (mguy-huntress) [inline firewall.go:139]: @rob-no We will not automatically commit a transaction.  That is up to the caller to do.  However, if we detect we have began a transaction without committing it, on Close, we will abort the transaction.

File: firewall.go (hunk 1 above)
Comment (mguy-huntress) [reply firewall.go:111]: We will abort an open transaction, see my comment below for details.

File: firewall.go (hunk 2 above)
Comment (mguy-huntress) [reply firewall.go:139]: The 4 functions in WFP that use transactions (Rules, SetRules, DeleteAll, and SaveRulesToDisk, all start their sessions in a transaction and call commit (unless given an override through the configuration task).
Here is an example of a commit being used:
https://github.com/huntresslabs/agent/blob/dbb8bcfe728b7a375f15852daa838c9f1c584b14/internal/hlib/wfp/wfp_windows.go#L54

Comment (rob-no) [reply firewall.go:139]: Ok I see that now. Feels a bit weird that the `beginTransaction` is abstracted away, but the commit isn't.

I think it's fine in this case, but in the future we shouldn't mix the pattern like that IMO

**Review by rob-no — APPROVED** [review]

File: firewall.go (hunk 2 above)
Comment (mguy-huntress) [reply firewall.go:139]: Just for additional info, it's a startup option to begin a transaction when the session is created.  You don't have to start a transaction at that point, but if you don't you'll miss some of the operations in the session create function.  You can tell it to start a session without starting a transaction and then call BeginTransaction (but as mentioned, then some of the WFP calls are not going to be a part of the transaction)

**Review by calebstewart — APPROVED** [review]
All looks good to me. I honestly didn't even realize that WFP supported transactions so this is great! :)

I couldn't run the unit tests on my machine, but I'm almost certain it's an issue with my Windows dev VM (which gets old and weird because I never use it). Since there's no CI/CD here yet apparently, I verified with @mguy-huntress that the unit tests pass appropriately. :+1:

**Pull request merged**